README.rst
XCRI-CAP to RDF and back again
==============================

This repository contains XSL stylesheets to convert XCRI-CAP 1.2 XML into RDF,
a Python script to turn such RDF back into XML, and SKOS concept schemes for a
few relevant controlled vocabularies.

Stylesheets
-----------

These stylesheets convert XCRI-CAP 1.2 into RDF/XML using appropriate
vocabularies.

You will probably want to extend and override parts of the main stylesheet
(``xcri2rdf.xsl``) in order to model locally-important concepts and provide
identifiers for your courses, presentations, and so forth.


Using
~~~~~

Make sure you have an XSLT 2.0 processor installed (we'll assume `saxon
<http://saxon.sourceforge.net/>`_).

Then, simply::

    $ saxon your-xcri-feed.xml xcri2rdf.xsl > your-xcri-feed.rdf


Overriding
~~~~~~~~~~

The ``daisy2rdf.xsl`` stylesheet should give you some clues. Note the ``<xsl:import />`` line, and how we have overriden the ``rdf-about`` template to check some special case before deferring to the implementation in ``xcri2rdf.xsl`` using ``<xsl:apply-imports />``.


Python script
-------------

Prerequisites
~~~~~~~~~~~~~

You'll need the ``rdflib`` module, which you can install using either ``easy_install`` or ``pip``, or using a Linux distribution package manager.

API
~~~

``python/xcri_rdf.py`` contains an ``XCRICAPSerializer`` class which takes an ``rdflib.Graph`` instance as its only argument. The graph must contain a subject of type ``xcri:catalog``::

    import rdflib
    from xcri_rdf import XCRICAPSerializer

    graph = rdflib.ConjunctiveGraph()
    graph.parse('course-data.rdf')
    serializer = XCRICAPSerializer(graph)

You can then serialize the graph to XML using either its ``serialize`` or ``generator`` methods. The former takes a file-like object (e.g. a file, or a socket), whereas the later yield strings::

    import sys
    serializer.serialize(sys.stdout)

    # or

    for str in serializer.generator():
        print str

As a script
~~~~~~~~~~~

The Python module may also be invoked as a script::

    $ python -m xcri_rdf file-one.rdf thesauri/*/*.rdf

The script is invoked with some number of RDF/XML files as arguments. Include the thesauri files so that things like studyMode and attendancePattern are suitably serialized.


Feedback
--------

Feedback is very welcome to `COURSEDATASTAGE1@JISCMAIL.AC.UK
<mailto:COURSEDATASTAGE1@JISCMAIL.AC.UK>`_.


debian/changelog
xcri-rdf (0.1) unstable; urgency=low

  * Initial packaging

 -- Alexander Dutton <alexander.dutton@it.ox.ac.uk>  Sat, 2 Feb 2013 21:06:07 +0000



tests/test.py
import os
import subprocess
import threading
import unittest

from lxml import etree
import rdflib

def relative_path(*args):
    return os.path.join(os.path.dirname(__file__), *args)

class Namespaces(dict):
    def __getattr__(self, name): return self[name]
NS = {'dc': 'http://purl.org/dc/elements/1.1/',
      'dcterms': 'http://purl.org/dc/terms/',
      'mlo': 'http://purl.org/net/mlo/',
      'xcri': 'http://xcri.org/profiles/1.2/',
      'xtypes': 'http://purl.org/xtypes/',
      'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
      'xhtml': 'http://www.w3.org/1999/xhtml#',
      'skos': 'http://www.w3.org/2004/02/skos/core#'}
XMLNS = NS.copy()
# Some namespaces differ between RDF and XML
XMLNS.update({'mlo': 'http://purl.org/net/mlo',
              'xcri': 'http://xcri.org/profiles/1.2/catalog',
              'xhtml': 'http://www.w3.org/1999/xhtml',
              'xsi': 'http://www.w3.org/2001/XMLSchema-instance'})
NS = Namespaces((k, rdflib.Namespace(NS[k])) for k in NS)

common_descriptive_elements = [(NS.dcterms.description, 'dc:description'),
                               (NS.xcri['abstract'], 'xcri:abstract'),
                               (NS.xcri.applicationProcedure, 'xcri:applicationProcedure'),
                               (NS.mlo.assessment, 'mlo:assessment'),
                               (NS.xcri.learningOutcome, 'xcri:learningOutcome'),
                               (NS.mlo.objective, 'mlo:objective'),
                               (NS.mlo.prerequisite, 'mlo:prerequisite'),
                               (NS.xcri.regulations, 'xcri:regulations')]

filenames = [('Conted', relative_path('data', 'conted.xcricap')),
             ('AdamSmith', relative_path('data', 'adamsmith.xcricap'))]

class XSLTTestCase(unittest.TestCase):
    _cache = threading.local()
    _cache.graph, _cache.xml = {}, {}
    filename = None

    bases = {'catalog': 'http://example.org/catalog/',
             'provider': 'http://example.org/provider/',
             'course': 'http://example.org/course/',
             'presentation': 'http://example.org/presentation/',
             'venue': 'http://example.org/venue/'}

    def assertXMLEqual(self, one, two, message=None):
        self.assertEqual(one.tag, two.tag, message)
        self.assertEqual(one.attrib, two.attrib, message)
        self.assertEqual(one.text, two.text, message)
        for children in zip(one, two):
            self.assertXMLEqual(children[0], children[1], message)
        self.assertEqual(one.tail, two.tail, message)

    def assertComplexEqual(self, rdf, xml, message=None):
        if isinstance(rdf, rdflib.URIRef):
            self.assertEqual(rdf, xml.attrib['href'], message)
        elif not isinstance(rdf, rdflib.Literal):
            raise AssertionError("Unexpected RDF term: %r" % rdf)
        elif rdf.datatype == NS.xtypes['Fragment-XHTML']:
            self.assertXMLEqual(etree.fromstring(rdf), xml[0], message)
        elif rdf.datatype == None:
            self.assertEqual(rdf, xml.text, message)
        else:
            raise AssertionError("Unexpected RDF datatype: %s" % rdf.datatype)

    def checkIdentifier(self, rdf, xml):
        if isinstance(rdf, rdflib.URIRef):
            if xml.xpath("dc:identifier[text()='{0}']".format(rdf), namespaces=XMLNS):
                return True
            elif xml.xpath("dc:identifier[not(@xsi:type) and text()='{0}{1}']".format(self.bases[xml.tag.rsplit('}')[-1]], rdf), namespaces=XMLNS):
                return True
            else:
                return False
        elif isinstance(rdf, rdflib.BNode):
            if xml.xpath("dc:identifier[not(@xsi:type) and text()='{0}{1}']".format(self.bases[xml.tag.rsplit('}')[-1]], rdf), namespaces=XMLNS):
                return False
        else:
            return False
        return True
    def assertValidIdentifier(self, rdf, xml):
        if not self.checkIdentifier(rdf, xml):
            raise AssertionError("{0} doesn't match expected identifier for {1}".format(rdf, xml))

    def assertPropertyEqual(self, node, predicate, xml, xpath):
        value = self.graph.value(node, predicate, any=False)
        elements = xml.xpath(xpath, namespaces=XMLNS)
        if value is not None:
            self.assertEqual(value, elements[0].text or '')
        else:
            self.assertEqual(len(elements), 0)
    
    def zipThings(self, nodes, elements):
        nodes = list(nodes)
        nodes.sort(key=lambda n: self.graph.value(n, NS.xcri['x-order']).toPython())
        return zip(nodes, elements)

    @property
    def saxon_path(self):
        candidates = ['/usr/bin/saxon', '/usr/bin/saxonb-xslt']
        for candidate in candidates:
            if os.path.exists(candidate):
                return candidate
        raise ImproperlyConfigured("Couldn't find saxon.")

    @property
    def graph(self):
        if not self.filename:
            raise NotImplementedError("filename not set")
        if not self._cache.graph.get(self.filename):
            saxon = subprocess.Popen([self.saxon_path,
                                      self.filename,
                                      relative_path('..', 'stylesheets', 'xcri2rdf.xsl'),
                                      "catalog-base="+self.bases['catalog'],
                                      "provider-base="+self.bases['provider'],
                                      "course-base="+self.bases['course'],
                                      "presentation-base="+self.bases['presentation'],
                                      "venue-base="+self.bases['venue'],
                                      "order-annotation=true"],
                                     stdout=subprocess.PIPE,
                                     stderr=open('/dev/null', 'w'))
            self._cache.graph[self.filename] = rdflib.ConjunctiveGraph()
            self._cache.graph[self.filename].parse(saxon.stdout)
        return self._cache.graph[self.filename]

    @property
    def xml(self):
        if not self._cache.xml.get(self.filename):
            with open(self.filename) as f:
                self._cache.xml[self.filename] = etree.parse(f)
        return self._cache.xml[self.filename]

    def testTransform(self):
        self.graph

    def testCatalog(self):
        graph, xml = self.graph, self.xml
        catalog = self.graph.value(None, NS.rdf.type, NS.xcri.catalog, any=False)
        xml_catalog = xml.xpath('/xcri:catalog', namespaces=XMLNS)[0]

        self.assertValidIdentifier(catalog, xml_catalog)

        # Check that catalog title and description were picked up correctly.
        self.assertPropertyEqual(catalog, NS.dcterms.title, xml_catalog, 'dc:title')
        self.assertPropertyEqual(catalog, NS.dcterms.description, xml_catalog, 'dc:description')

        # Check that there were courses to serialize, and that we've
        # serialized the right number.
        self.assert_(len(xml_catalog.xpath('xcri:provider/xcri:course', namespaces=XMLNS)))
        self.assertEqual(len(list(graph.objects(catalog, NS.skos.member))),
                         len(xml_catalog.xpath('xcri:provider/xcri:course', namespaces=XMLNS)))

    def testProvider(self):
        graph = self.graph
        providers = list(graph.subjects(NS.rdf.type, NS.xcri.provider))

        # There should be at least one provider
        self.assert_(len(providers) > 0)

        for provider, xml_provider in self.zipThings(providers, self.xml.xpath('/xcri:catalog/xcri:provider', namespaces=XMLNS)):

            # Check that the provider's name and description made it across
            self.assertPropertyEqual(provider, NS.dcterms.title, xml_provider, 'dc:title')
            self.assertPropertyEqual(provider, NS.dcterms.description, xml_provider, 'dc:description')

    def testCourse(self):
        graph, xml = self.graph, self.xml
        courses = list(graph.subjects(NS.rdf.type, NS.xcri.course))

        # There should be at least one course
        self.assert_(len(courses) > 0)


        for course, xml_course in self.zipThings(courses, self.xml.xpath('.//xcri:course', namespaces=XMLNS)):

            # Check title and description
            self.assertEqual(graph.value(course, NS.dcterms['title']),
                             xml_course.xpath('dc:title', namespaces=XMLNS)[0].text)

            for predicate, element in common_descriptive_elements:
                rdf_cde = list(graph.objects(course, predicate))
                xml_cde = xml_course.xpath(element, namespaces=XMLNS)
                self.assertEqual(len(rdf_cde), len(xml_cde), "Mismatched {0} for {1} ({2} and {3}, {4})".format(element, course, len(rdf_cde), len(xml_cde), graph.value(course, NS.dcterms.identifier)))
                if len(rdf_cde) == 1:
                    self.assertComplexEqual(rdf_cde[0], xml_cde[0])
                elif len(rdf_cde) > 1:
                    raise AssertionError("More than one descriptive element: {0}, {1}".format(element, graph.value(course, NS.dcterms.identifier)))

def load_tests(loader, tests, pattern):
    suite = unittest.TestSuite()
    for name, filename in filenames:
        test_class = type('{0}TestCase'.format(name),
                          (XSLTTestCase,),
                          {'filename': filename})
        tests = loader.loadTestsFromTestCase(test_class)
        suite.addTests(tests)
    return suite


if __name__ == '__main__':
    unittest.main()


thesauri/jacs/generate_jacs.py
import csv
import sys

import rdflib

BASE = rdflib.URIRef('http://jacs.dataincubator.org/')
SKOS = rdflib.Namespace('http://www.w3.org/2004/02/skos/core#')
RDF = rdflib.Namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#')
notation = rdflib.URIRef(BASE + 'notation')

def main(rows):
    # Skip the first two rows
    rows.next()
    rows.next()

    yield (BASE, RDF.type, SKOS.ConceptScheme)
    yield (BASE, SKOS.prefLabel, rdflib.Literal('JACS codes'))
    for row in rows:
        code, label, definition = row[:3]
        uri = rdflib.URIRef(BASE + code.lower())
        if len(code) == 1:
            yield (BASE, SKOS.hasTopConcept, uri)
        else:
            yield (rdflib.URIRef(BASE + code.lower()[0]), SKOS.narrower, uri)
        yield (uri, RDF.type, SKOS.Concept)
        yield (uri, SKOS.prefLabel, rdflib.Literal(label))
        yield (uri, SKOS.notation, rdflib.Literal(code, datatype=notation))
        if definition:
            yield (uri, SKOS.definition, rdflib.Literal(definition))

if __name__ == '__main__':
    graph = rdflib.ConjunctiveGraph()
    with open('jacs.csv') as f:
        for triple in main(csv.reader(f)):
            graph.add(triple)
    graph.serialize(sys.stdout, format='pretty-xml')


python/xcri_rdf.py
import collections
import datetime
import itertools
import re
import types
try:
    from cStringIO import StringIO
except ImportError:
    import StringIO
from xml.sax.saxutils import XMLGenerator

import dateutil.parser
from lxml import etree
import rdflib

XMLNS = {
    'xcri': 'http://xcri.org/profiles/1.2/catalog',
    'mlo': 'http://purl.org/net/mlo',
    'dc': 'http://purl.org/dc/elements/1.1/',
    'xsi': 'http://www.w3.org/2001/XMLSchema-instance',
    'cdp': 'http://xcri.co.uk',
    'html': 'http://www.w3.org/1999/xhtml',
}
INVERSE_XMLNS = tuple((v, k) for k, v in sorted(XMLNS.items(), key=lambda (k,v): -len(v)))

NS = {
    'skos': 'http://www.w3.org/2004/02/skos/core#',
    'foaf': 'http://xmlns.com/foaf/0.1/',
#    'xcri': 'http://xcri.org/profiles/1.2/catalog/',
    'xcri': 'http://xcri.org/profiles/1.2/',
    'rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
    'rdfs': 'http://www.w3.org/2000/01/rdf-schema#',
    'dc': 'http://purl.org/dc/elements/1.1/',
    'dcterms': 'http://purl.org/dc/terms/',
    'spatialrelations': 'http://data.ordnancesurvey.co.uk/ontology/spatialrelations/',
    'mlo': 'http://purl.org/net/mlo/',
    'time': 'http://www.w3.org/2006/time#',
    'xtypes': 'http://purl.org/xtypes/',
    'v': 'http://www.w3.org/2006/vcard/ns#',
    'xsd': 'http://www.w3.org/2001/XMLSchema#',
}
class _NS(dict):
    def __init__(self, ns):
        super(_NS, self).__init__((prefix, rdflib.Namespace(uri)) for prefix, uri in ns.iteritems())
    def __getattr__(self, name):
        return self[name]
NS = _NS(NS)

is_localpart = re.compile(u"""^[A-Z _ a-z \xc0-\xd6 \xd8-\xf6 \xf8-\xff \u037f-\u1fff \u200c-\u218f]
                               [A-Z _ a-z \xc0-\xd6 \xd8-\xf6 \xf8-\xff \u037f-\u1fff \u200c-\u218f \\- . \\d]*$""",
                          re.VERBOSE).match

def xsi_type(uri):
    for ns, prefix in INVERSE_XMLNS:
        if uri.startswith(ns):
            localpart = uri[len(ns):]
            if is_localpart(localpart):
                return {'xsi:type': "%s:%s" % (prefix, localpart)}
    for i in xrange(len(uri)):
        if is_localpart(uri[i:]):
            return {'xsi:type' :'ns:%s' % uri[i:], 'xmlns:ns': uri[:i]}
    return {}

identifiers = (NS.dc.identifier, NS.dcterms.identifier)
subject_predicates = (NS.dcterms.subject, NS.dc.subject)
labels = (NS.skos.prefLabel, NS.rdfs.label, NS.dcterms['title'], NS.dc['title'])
urls = (NS.foaf.homepage, NS.foaf.page)
descriptions = (NS.dcterms.description, NS.rdfs.comment)

def _find_first(name, ps, rich_element=False):
    def f(self, xg, entity):
        for obj in itertools.chain(*(self.graph.objects(entity, p) for p in ps)):
            if rich_element:
                self.descriptive_text_element(xg, name, obj)
            else:
                xg.textualElement(name, {}, unicode(obj))
            break
    return f

class IndentingXMLGenerator(XMLGenerator):
    def __init__(self, *args, **kwargs):
        XMLGenerator.__init__(self, *args, **kwargs)
        self.depth = 0
        self.chars = False

    def startElement(self, name, attrs):
        if self.depth and not self.chars:
            XMLGenerator.characters(self, '\n' + '  '*self.depth)
        XMLGenerator.startElement(self, name, attrs)
        self.depth += 1
        self.chars = False

    def endElement(self, name):
        self.depth -= 1
        if not self.chars:
            self.characters('\n' + '  '*self.depth)
        XMLGenerator.endElement(self, name)
        self.chars = False

    def characters(self, content):
        XMLGenerator.characters(self, content)
        self.chars = True

    def textualElement(self, name, attrs, content):
        self.startElement(name, attrs)
        self.characters(content)
        self.endElement(name)

def serialize_etree(xml, xg, previous_nsmap={}):
    attrib = dict(xml.attrib)
    for ns in xml.nsmap:
        if xml.nsmap[ns] == previous_nsmap.get(ns):
            continue
        if ns is None:
            attrib['xmlns'] = xml.nsmap[ns]
        else:
            attrib['xmlns:%s' % ns] = xml.nsmap[ns]
    if xml.prefix is None:
        tag = xml.tag.split('}', 1)[-1]
    else:
        tag = '%s:%s' % (xml.prefix, xml.tag.split('}', 1)[1])
    xg.startElement(tag, attrib)

    if xml.text:
        xg.characters(xml.text)
    for child in xml:
        serialize_etree(child, xg, xml.nsmap)
        if child.tail:
            xg.characters(child.tail)
    xg.endElement(tag)

class XCRICAPSerializer(object):
    common_descriptive_elements = [
        (NS.xcri['abstract'], 'xcri:abstract'),
        (NS.xcri.applicationProcedure, 'xcri:applicationProcedure'),
        (NS.mlo.assessment, 'mlo:assessment'),
        (NS.xcri.learningOutcome, 'xcri:learningOutcome'),
        (NS.mlo.objective, 'mlo:objective'),
        (NS.mlo.prerequisite, 'mlo:prerequisite'),
        (NS.xcri.regulations, 'xcri:regulations'),
    ]

    address_elements = [
        (NS.v['street-address'], 'mlo:address'),
        (NS.v['extended-address'], 'mlo:address'),
        (NS.v['locality'], 'mlo:address'),
        (NS.v['postal-code'], 'mlo:postcode'),
    ]

    xsi_schema_locations = {
        "http://xcri.org/profiles/1.2/catalog": "http://www.xcri.co.uk/bindings/xcri_cap_1_2.xsd",
        "http://xcri.org/profiles/1.2/catalog/terms": "http://www.xcri.co.uk/bindings/xcri_cap_terms_1_2.xsd",
        "http://xcri.co.uk": "http://www.xcri.co.uk/bindings/coursedataprogramme.xsd",
        "http://www.w3.org/2003/01/geo/wgs84_pos": "http://xcri-dev.conted.ox.ac.uk/geo.xsd",
    }

    controlled_vocabularies = [
        (NS.xcri.studyMode, 'xcri:studyMode', rdflib.URIRef('http://xcri.org/profiles/catalog/1.2/studyMode/notation')),
        (NS.xcri.attendanceMode, 'xcri:attendanceMode', rdflib.URIRef('http://xcri.org/profiles/catalog/1.2/attendanceMode/notation')),
        (NS.xcri.attendancePattern, 'xcri:attendancePattern', rdflib.URIRef('http://xcri.org/profiles/catalog/1.2/attendancePattern/notation')),
    ]
    
    xmlns = XMLNS.copy()

    def __init__(self, graph, catalog=None, encoding='utf-8', simple=True):
        self.graph = graph
        self.catalog = catalog or self.graph.value(None, NS.rdf.type, NS.xcri.catalog, any=False)
        self.encoding = encoding
        self.simple = simple

    def serialize(self, stream):
        self.stream = stream
        stack = [self._serialize()]
        while stack:
            try:
                item = stack[-1].next()
            except StopIteration:
                stack.pop()
            else:
                if isinstance(item, types.GeneratorType):
                    stack.append(item)

    def generator(self):
        self.stream = StringIO()
        stack = [self._serialize()]
        while stack:
            try:
                item = stack[-1].next()
            except StopIteration:
                stack.pop()
            else:
                if isinstance(item, types.GeneratorType):
                    stack.append(item)
            yield self.stream.getvalue()
            self.stream.seek(0)
            self.stream.truncate()
        yield self.stream.getvalue()
    
    def _serialize(self):
        xg = IndentingXMLGenerator(self.stream, self.encoding)
        xg.startDocument()
        yield self.catalog_element(xg, self.catalog)
        xg.endDocument()

    def catalog_element(self, xg, catalog):
        attrib = {}
        attrib.update(('xmlns:%s' % prefix, uri) for prefix, uri in self.xmlns.iteritems())
        attrib['xsi:schemaLocation'] = ' '.join(map(' '.join, self.xsi_schema_locations.iteritems()))
        attrib['generated'] = datetime.datetime.utcnow().isoformat() + '+00:00'

        xg.startElement('xcri:catalog', attrib)
        yield self.catalog_content(xg, catalog)
        xg.endElement('xcri:catalog')
    
    def catalog_content(self, xg, catalog):
        yield self.serialize_common(xg, catalog)

        provider_courses, catalogs = collections.defaultdict(set), set([catalog])
        while catalogs:
            for member in self.graph.objects(catalogs.pop(), NS.skos.member):
                if (member, NS.rdf.type, NS.xcri.course) in self.graph:
                    provider = self.graph.value(None, NS.mlo.offers, member)
                    provider_courses[provider].add(member)
                elif (member, NS.rdf.type, NS.xcri.catalog) in self.graph:
                    catalogs.add(member)

        if self.simple:
            provider = self.graph.value(catalog, NS.dcterms.publisher)
            yield self.provider_element(xg,
                                        provider,
                                        itertools.chain(*provider_courses.itervalues()))
        else:
            for provider, courses in provider_courses.iteritems():
                yield self.provider_element(xg, provider, courses)

    def provider_element(self, xg, provider, courses):
        xg.startElement('xcri:provider', {})
        yield self.provider_content(xg, provider, courses)
        xg.endElement('xcri:provider')

    def provider_content(self, xg, provider, courses):
        if provider:
            yield self.serialize_common(xg, provider)
            yield self.serialize_location(xg, provider)
        for course in courses:
            yield self.course_element(xg, course)

    def course_element(self, xg, course):
        xg.startElement('xcri:course', {})
        yield self.course_content(xg, course)
        xg.endElement('xcri:course')

    def course_content(self, xg, course):
        yield self.serialize_common(xg, course)
        yield self.serialize_common_descriptive_elements(xg, course)
        yield self.serialize_subjects(xg, course)
        for presentation in self.graph.objects(course, NS.mlo.specifies):
            yield self.presentation_element(xg, presentation)

    def presentation_element(self, xg, presentation):
        xg.startElement('xcri:presentation', {})
        yield self.presentation_content(xg, presentation)
        xg.endElement('xcri:presentation')

    def presentation_content(self, xg, presentation):
        self.serialize_common(xg, presentation)
        self.serialize_common_descriptive_elements(xg, presentation)
        self.serialize_date(xg, presentation, NS.mlo.start, 'mlo:start')
        self.serialize_date(xg, presentation, NS.xcri.end, 'xcri:end')
        self.serialize_date(xg, presentation, NS.xcri.applyFrom, 'xcri:applyFrom')
        self.serialize_date(xg, presentation, NS.xcri.applyUntil, 'xcri:applyUntil')
        self.serialize_places(xg, presentation)
        self.serialize_applyTo(xg, presentation)
        self.serialize_controlled_vocabularies(xg, presentation)
        for venue in self.graph.objects(presentation, NS.xcri.venue):
            yield self.venue_element(xg, venue)

    def venue_element(self, xg, venue):
        """
        Serializes venue information if there is any.
        """

        xg.startElement('xcri:venue', {})
        xg.startElement('xcri:provider', {})
        yield self.serialize_common(xg, venue)
        yield self.serialize_location(xg, venue)
        xg.endElement('xcri:provider')
        xg.endElement('xcri:venue')

    def serialize_common(self, xg, entity):
        self.serialize_identifiers(xg, entity)
        self.serialize_title(xg, entity)
        self.serialize_url(xg, entity)
        self.serialize_description(xg, entity)

    def serialize_location(self, xg, entity):
        with_address, address = entity, None
        while with_address:
            address = self.graph.value(entity, NS.v.adr)
            if address:
                break
            with_address = self.graph.value(with_address, NS.spatialrelations.within)
        else:
            return

        if any(self.graph.value(address, prop) for prop, name in self.address_elements):
            xg.startElement('mlo:location', {})
            for prop, name in self.address_elements:
                obj = self.graph.value(address, prop)
                if obj:
                    xg.textualElement(name, {}, unicode(obj))
            xg.endElement('mlo:location')

    def serialize_controlled_vocabularies(self, xg, presentation):
        for prop, name, datatype in self.controlled_vocabularies:
            value = self.graph.value(presentation, prop)
            if not value:
                continue
            # Only use notations with the correct datatype (i.e., scheme)
            for notation in self.graph.objects(value, NS.skos.notation):
                if notation.datatype == datatype:
                    break
            else:
                notation = None
            for label in labels:
                content = self.graph.value(value, label)
                if content:
                    break
            if not (notation or content):
                continue
            xg.textualElement(name,
                              {'identifier': unicode(notation)} if notation else {},
                              content or '')


    def serialize_subjects(self, xg, course):
        subjects = set(itertools.chain(*(self.graph.objects(course, p) for p in subject_predicates)))
        for subject in set(subjects):
            subjects.update(self.graph.subjects(NS.skos.narrower, subject))
            subjects.update(self.graph.objects(subject, NS.skos.broader))
        for subject in set(subjects):
            subjects.update(self.graph.objects(subject, NS.skos.related))
        
        for subject in subjects:
            if isinstance(subject, rdflib.Literal):
                attrib, content = {}, unicode(subject)
            else:
                notation = self.graph.value(subject, NS.skos.notation)
                attrib = xsi_type(notation.datatype) if notation else {}
                if notation:
                    attrib['identifier'] = unicode(notation)
                for label in labels:
                    content = self.graph.value(subject, label)
                    if content:
                        break
                # The non-RDF people do this one specially.
                if subject.startswith('http://jacs.dataincubator.org/'):
                    attrib['xsi:type'] = 'cdp:JACS3'
            if attrib or content:
                xg.textualElement('dc:subject', attrib, content or '')


    def serialize_date(self, xg, entity, prop, name):
        dt = self.graph.value(entity, prop)
        if isinstance(dt, rdflib.Literal):
            dtf, content = dt.toPython(), None
            if isinstance(dtf, basestring):
                dtf, content = None, dtf
            if not isinstance(dtf, (datetime.datetime, datetime.date)):
                dtf = None
        else:
            dtf = self.graph.value(dt, NS.time.inXSDDateTime) or self.graph.value(dt, NS.rdf.value)
            for label in labels:
                content = self.graph.value(dt, label)
                if content:
                    break
        if not (dtf or content):
            return
        if not content:
            try:
                parsed = dateutil.parser.parse(dtf)
            except ValueError:
                content = dtf
            else:
                if dtf.datatype == NS.xsd.date:
                    content = parsed.strftime("%A, %d %B %Y")
                elif dtf.datatype == NS.xsd.dateTime:
                    content = parsed.strftime("%I:%M %p, %A, %d %B %Y")
        attrib = {'dtf': dtf} if dtf else {}
        xg.textualElement(name, attrib, content or '')

    def serialize_identifiers(self, xg, entity):
        for obj in self.graph.objects(entity, NS.skos.notation):
            attrib = xsi_type(obj.datatype)
            if not attrib:
                continue
            xg.textualElement('dc:identifier', attrib, unicode(obj))

        plain_identifiers = set()
        if isinstance(entity, rdflib.URIRef):
            plain_identifiers.add(unicode(entity))
        for obj in itertools.chain(*(self.graph.objects(entity, p) for p in identifiers)):
            if isinstance(obj, rdflib.Literal):
                plain_identifiers.add(unicode(obj))
        for identifier in plain_identifiers:
            xg.textualElement('dc:identifier', {}, identifier)

    def serialize_common_descriptive_elements(self, xg, entity):
        for prop, name in self.common_descriptive_elements:
            for obj in self.graph.objects(entity, prop):
                self.descriptive_text_element(xg, name, obj)

    def descriptive_text_element(self, xg, name, obj, attrib={}):
        """
        Serializes RDF terms that are either links, HTML or plain text.
        """
        if isinstance(obj, rdflib.URIRef):
            xg.startElement(name, {'href': unicode(obj)})
            xg.endElement(name)
        else:
            if obj.datatype == NS.xtypes['Fragment-XHTML']:
                xml = etree.fromstring(unicode(obj))
            elif obj.datatype == NS.xtypes['Fragment-HTML']:
                xml = etree.fromstring(unicode(obj), parser=etree.HTMLParser())
            else:
                xg.textualElement(name, attrib, unicode(obj))
                return
            xg.startElement(name, attrib)
            serialize_etree(xml, xg, self.xmlns)
            xg.endElement(name)

    serialize_title = _find_first('dc:title', labels)
    serialize_url = _find_first('mlo:url', urls)
    serialize_description = _find_first('dc:description', descriptions, rich_element=True)
    serialize_places = _find_first('mlo:places', (NS.mlo.places,))
    serialize_applyTo = _find_first('xcri:applyTo', (NS.xcri.applyTo,))

   


if __name__ == '__main__':
    import sys

    graph = rdflib.ConjunctiveGraph()
    if len(sys.argv) > 1:
        for name in sys.argv[1:]:
            with open(name, 'r') as f:
                graph.parse(f)
    else:
        graph.parse(sys.stdin)

    serializer = XCRICAPSerializer(graph)
    serializer.serialize(sys.stdout)
